src/FirmwarePlugin/APM/APMFirmwarePlugin.cc

```diff
@@ -167,5 +251,66 @@ void APMFirmwarePlugin::adjustMavlinkMessage(mavlink_message_t* message)
         mavlink_msg_param_set_encode(message->sysid, message->compid, message, &paramSet);
     }
 
-    // FIXME: Need to implement mavlink message severity adjustment
+    if (message->msgid == MAVLINK_MSG_ID_STATUSTEXT)
+    {
+        QByteArray b;
+        b.resize(MAVLINK_MSG_STATUSTEXT_FIELD_TEXT_LEN+1);
+        mavlink_msg_statustext_get_text(message, b.data());
+        // Ensure NUL-termination
+        b[b.length()-1] = '\0';
+        QString text = QString(b);
+        qCDebug(FirmwareUpgradeLog()) << text;
+
+        if (text.contains(APM_COPTER_REXP) || text.contains(APM_PLANE_REXP) || text.contains(APM_ROVER_REXP)) {
+            // found version string
+            _firmwareVersion = APMFirmwareVersion(text);
+        }
+
+        // adjust mesasge if needed
+        _adjustSeverity(message);
+    }
+}
+
+void APMFirmwarePlugin::_adjustSeverity(mavlink_message_t* message) const
+{
+    // return if we don't know the firmware version
+    if (!_firmwareVersion.isValid()) {
+        return;
+    }
+
+    bool adjustmentNeeded = false;
+    if (_firmwareVersion.vehicleType().contains(APM_COPTER_REXP)) {
+        if (_firmwareVersion < APMFirmwareVersion(MIN_COPTER_VERSION_WITH_CORRECT_SEVERITY_MSGS)) {
+            adjustmentNeeded = true;
+        }
+    } else if (_firmwareVersion.vehicleType().contains(APM_PLANE_REXP)) {
+        if (_firmwareVersion < APMFirmwareVersion(MIN_PLANE_VERSION_WITH_CORRECT_SEVERITY_MSGS)) {
+            adjustmentNeeded = true;
+        }
+    } else if (_firmwareVersion.vehicleType().contains(APM_ROVER_REXP)) {
+        if (_firmwareVersion < APMFirmwareVersion(MIN_ROVER_VERSION_WITH_CORRECT_SEVERITY_MSGS)) {
+            adjustmentNeeded = true;
+        }
+    }
+
+    if (!adjustmentNeeded) {
+        return;
+    }
+
+    // finally lets make QGC happy with right severity values
+    mavlink_statustext_t statusText;
+    mavlink_msg_statustext_decode(message, &statusText);
+    switch(statusText.severity) {
+    case MAV_SEVERITY_ALERT:    /* SEVERITY_LOW according to old codes */
```

Comment:
I am done with the review comments. Pushed new changes.

I am undecided about the rover version. But I guess we will have to wait
for what APM team decides. If they go by the current decision of releasing
it as 2.6.0 this code will fail for rover, I will put in the hack lateron
to consider 2.6.0 bigger than 2.50 if they don't change their mind.

On Fri, Oct 9, 2015 at 5:32 AM, Don Gagne <notifications@github.com> wrote:

> In src/FirmwarePlugin/APM/APMFirmwarePlugin.cc
> <https://github.com/mavlink/qgroundcontrol/pull/1926#discussion_r41585293>
> :
>
> > +        }
> > +    } else if (_firmwareVersion.vehicleType().contains(APM_ROVER_REXP)) {
> > +        if (_firmwareVersion < APMFirmwareVersion(MIN_ROVER_VERSION_WITH_CORRECT_SEVERITY_MSGS)) {
> > +            adjustmentNeeded = true;
> > +        }
> > +    }
> > +
> > +    if (!adjustmentNeeded) {
> > +        return;
> > +    }
> > +
> > +    // finally lets make QGC happy with right severity values
> > +    mavlink_statustext_t statusText;
> > +    mavlink_msg_statustext_decode(message, &statusText);
> > +    switch(statusText.severity) {
> > +    case MAV_SEVERITY_ALERT:    /* SEVERITY_LOW according to old codes */
>
> The coding style has the interior of a switch indented 4 spaces.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mavlink/qgroundcontrol/pull/1926/files#r41585293>.
>



-- 
Regards,
pritam